File: android/src/main/kotlin/com/wultra/android/powerauth/flutter/internal/services/PowerAuthEncryptorService.kt
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wultra.android.powerauth.flutter.internal.services

import android.content.Context
import com.wultra.android.powerauth.flutter.Constants
import com.wultra.android.powerauth.flutter.Errors
import com.wultra.android.powerauth.flutter.ManagedAny
import com.wultra.android.powerauth.flutter.PowerAuthObjectRegister
import com.wultra.android.powerauth.flutter.ReleasePolicy
import com.wultra.android.powerauth.flutter.WrapperException
import com.wultra.android.powerauth.flutter.internal.core.BasePowerAuthService
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.getlime.security.powerauth.core.CoreEncryptedResponse
import io.getlime.security.powerauth.core.CoreEncryptor
import io.getlime.security.powerauth.networking.response.IGetEncryptorListener
import io.getlime.security.powerauth.sdk.PowerAuthSDK

internal class PowerAuthEncryptorService(
    private val objectRegister: PowerAuthObjectRegister,
    @Suppress("UNUSED_PARAMETER") context: Context
) : BasePowerAuthService(objectRegister) {

    override val name = "encryptor"

    private companion object ArgKeys {
        const val SCOPE = "scope"
        const val INSTANCE_ID = "powerAuthInstanceId"
        const val OBJECT_ID = "objectId"
        const val REQUEST_BODY = "requestBody"
        const val RESPONSE_BODY = "responseBody"
    }

    private object HandlerNames {
        const val INITIALIZE = "initialize"
        const val CAN_ENCRYPT_REQUEST = "canEncryptRequest"
        const val ENCRYPT_REQUEST = "encryptRequest"
        const val CAN_DECRYPT_RESPONSE = "canDecryptResponse"
        const val DECRYPT_RESPONSE = "decryptResponse"
    }

    override val handlers by lazy {
        mapOf(
            HandlerNames.INITIALIZE to this::initialize,
            HandlerNames.CAN_ENCRYPT_REQUEST to this::canEncryptRequest,
            HandlerNames.ENCRYPT_REQUEST to this::encryptRequest,
            HandlerNames.CAN_DECRYPT_RESPONSE to this::canDecryptResponse,
            HandlerNames.DECRYPT_RESPONSE to this::decryptResponse
        )
    }

    private fun initialize(call: MethodCall, result: Result) {
        try {
            val scope: String = call.getRequiredArgument(SCOPE)
            val instanceId: String = call.getRequiredArgument(INSTANCE_ID)
            val isActivationScope = when (scope) {
                "application" -> false
                "activation" -> true
                else -> throw WrapperException(
                    Errors.EC_WRONG_PARAMETER,
                    "Unknown scope value: $scope"
                )
            }

            val sdk = objectRegister.findObject(instanceId, PowerAuthSDK::class.java)
                ?: throw WrapperException(
                    Errors.EC_INSTANCE_NOT_CONFIGURED,
                    "PowerAuth instance '$instanceId' not configured."
                )

            val listener = object : IGetEncryptorListener {
                override fun onGetEncryptorSuccess(encryptor: CoreEncryptor) {
                    try {
                        //check if sdk has not been deconfigured in the meantime
                        val objectId = objectRegister.registerObjectIfOwnerMatches(
                            instanceId,
                            sdk,
                            ManagedAny.wrap(encryptor) { it.destroy() },
                            listOf(ReleasePolicy.keepAlive(Constants.ENCRYPTOR_KEEP_ALIVE_TIME))
                        ) ?: throw WrapperException(
                            Errors.EC_INSTANCE_NOT_CONFIGURED,
                            "PowerAuth instance '$instanceId' not configured."
                        )
                        result.success(objectId)
                    } catch (t: Throwable) {
                        encryptor.destroy()
                        Errors.error(result, t)
                    }
                }

                override fun onGetEncryptorFailed(t: Throwable) {
                    Errors.error(result, t)
                }
            }

            if (isActivationScope) {
                sdk.getEncryptorForActivationScope(listener)
            } else {
                sdk.getEncryptorForApplicationScope(listener)
            }
        } catch (t: Throwable) {
            Errors.error(result, t)
        }
    }

    private fun canEncryptRequest(call: MethodCall, result: Result) {
        withEncryptor(call, result) { encryptor ->
            encryptor.canEncryptRequest()
        }
    }

    private fun encryptRequest(call: MethodCall, result: Result) {
        withEncryptor(call, result) { encryptor ->
            val requestBody: ByteArray? = call.argument(REQUEST_BODY)
            val encryptedRequest = encryptor.encryptRequest(requestBody)

            mapOf(
                "requestBody" to encryptedRequest.requestBody,
                "requestHeaders" to encryptedRequest.requestHeaders.map { header ->
                    mapOf("name" to header.key, "value" to header.value)
                }
            )
        }
    }

    private fun canDecryptResponse(call: MethodCall, result: Result) {
        withEncryptor(call, result) { encryptor ->
            encryptor.canDecryptResponse()
        }
    }

    private fun decryptResponse(call: MethodCall, result: Result) {
        val responseBody: ByteArray = try {
            call.getRequiredArgument(RESPONSE_BODY)
        } catch (t: Throwable) {
            Errors.error(result, t)
            return
        }

        withEncryptor(call, result, destroyAfter = true) { encryptor ->
            encryptor.decryptResponse(CoreEncryptedResponse(responseBody))
        }
    }

    private fun <T : Any> withEncryptor(
        call: MethodCall,
        result: Result,
        destroyAfter: Boolean = false,
        block: (CoreEncryptor) -> T
    ) {
        val objectId: String = call.getRequiredArgument(OBJECT_ID)
        try {
            val value = objectRegister.useObjectAndTransform(
                objectId,
                CoreEncryptor::class.java
            ) { encryptor ->
                block(encryptor)

            } ?: throw WrapperException(
                Errors.EC_INVALID_NATIVE_OBJECT,
                "Encryptor object '$objectId' is no longer valid."
            )
            result.success(value)
        } catch (t: Throwable) {
            Errors.error(result, t)
        } finally {
            if (destroyAfter) {
                objectRegister.removeObject(objectId, CoreEncryptor::class.java)
            }
        }
    }

}

File: ios/flutter_powerauth_mobile_sdk_plugin/Sources/flutter_powerauth_mobile_sdk_plugin/internal/services/PowerAuthEncryptorService.swift
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PowerAuth2
import Flutter
import Foundation

class PowerAuthEncryptorService: PowerAuthFlutterService {
    
    let name = "PowerAuthEncryptorService"
    private let register: PowerAuthObjectRegister
    
    init(register: PowerAuthObjectRegister) {
        self.register = register
    }
    
    let handlers = [
        "encryptor_initialize": initialize,
        "encryptor_canEncryptRequest": canEncryptRequest,
        "encryptor_encryptRequest": encryptRequest,
        "encryptor_canDecryptResponse": canDecryptResponse,
        "encryptor_decryptResponse": decryptResponse,
    ]
    
    fileprivate enum Args: String {
        case scope
        case powerAuthInstanceId
        case objectId
        case requestBody
        case responseBody
    }
    
    private func initialize(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
        let scope: String = try call.requireParameter(Args.scope)
        let instanceId: String = try call.requireParameter(Args.powerAuthInstanceId)
        let sdk = try register.requirePowerAuthSDK(id: instanceId)

        let callback: (PowerAuthEncryptor?, Error?) -> Void = { encryptor, error in
            Utils.wrapThrowBlock(result: result) {
                if let error {
                    throw error
                }
                guard let encryptor else {
                    throw PluginException(.unknownError, message: "PowerAuth SDK returned neither an encryptor nor an error.")
                }
                guard let objectId = self.register.add(
                    object: encryptor,
                    ifOwnerMatches: sdk,
                    ownerId: instanceId,
                    policies: [.keepAlive(Constants.ENCRYPTOR_KEEP_ALIVE_TIME)]
                ) else {
                    throw PluginException(.instanceNotConfigured, message: "PowerAuth instance is no longer configured.")
                }
                result(objectId)
            }
        }

        switch scope {
        case "application":
            _ = sdk.encryptorForApplicationScope(callback: callback)
        case "activation":
            _ = sdk.encryptorForActivationScope(callback: callback)
        default:
            throw PluginException(.wrongParameter, message: "Unknown scope value: \(scope)")
        }
    }
    
    private func canEncryptRequest(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
        result(try encryptor(call, touch: true).canEncryptRequest)
    }
    
    private func encryptRequest(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
        let encryptor = try encryptor(call, touch: true)
        let requestBody = call.optionalDataParameter(Args.requestBody)
        let encryptedRequest = try encryptor.encryptRequest(requestBody)
        result([
            "requestBody": Data(encryptedRequest.requestBody),
            "requestHeaders": encryptedRequest.requestHeaders.map(\.serializable)
        ])
    }
    
    private func canDecryptResponse(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
        result(try encryptor(call, touch: true).canDecryptResponse)
    }
    
    private func decryptResponse(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
        let objectId: String = try call.requireParameter(Args.objectId)
        defer {
            register.removeAny(id: objectId)
        }
        guard let encryptor: PowerAuthEncryptor = register.touch(id: objectId) else {
            throw PluginException(.invalidNativeObject, message: "Encryptor object is no longer valid.")
        }
        let responseBody = try call.requiredDataParameter(Args.responseBody)
        let encryptedResponse = try PowerAuthEncryptedResponse(responseBody: responseBody)
        result(try encryptor.decryptResponse(encryptedResponse))
    }
    
    private func encryptor(_ call: FlutterMethodCall, touch: Bool) throws -> PowerAuthEncryptor {
        let objectId: String = try call.requireParameter(Args.objectId)
        let encryptor: PowerAuthEncryptor? = touch
            ? register.touch(id: objectId)
            : register.find(id: objectId)
        guard let encryptor else {
            throw PluginException(.invalidNativeObject, message: "Encryptor object is no longer valid.")
        }
        return encryptor
    }
}
